src/pybind/mgr/volumes/module.py

```diff
@@ -339,3 +454,126 @@ def _cmd_fs_volume_ls(self, inbuf, cmd):
             })
 
         return 0, json.dumps(result, indent=2), ""
+
+    def _cmd_fs_subvolume_group_ls(self, inbuf, cmd):
+        vol_name = cmd['vol_name']
+
+        if not self._volume_exists(vol_name):
+            return -errno.ENOENT, "", \
+                   "Volume '{0}' not found".format(vol_name)
+
+        with SubvolumeClient(self, fs_name=vol_name) as svc:
+            result = svc.list_subvolume_groups()
+
+        return 0, json.dumps(result, indent=2), ""
+
+    def _cmd_fs_subvolume_ls(self, inbuf, cmd):
+        vol_name = cmd['vol_name']
+        group_name = cmd.get('group_name', None)
+
+        if not self._volume_exists(vol_name):
+            return -errno.ENOENT, "", \
+                   "Volume '{0}' not found".format(vol_name)
+
+        with SubvolumeClient(self, fs_name=vol_name) as svc:
+            if group_name and not svc.get_group_path(group_name):
+                return -errno.ENOENT, "", \
+                    "Subvolume group '{0}' not found".format(group_name)
+            result = svc.list_subvolumes(group_name)
+
+        return 0, json.dumps(result, indent=2), ""
+
+    def _cmd_fs_subvolume_getpath(self, inbuf, cmd):
+        vol_name = cmd['vol_name']
+        sub_name = cmd['sub_name']
+
+        group_name = cmd.get('group_name', None)
+
+        if not self._volume_exists(vol_name):
+            return -errno.ENOENT, "", "Volume '{0}' not found".format(vol_name)
+
+        with SubvolumeClient(self, fs_name=vol_name) as svc:
```

Comment:
I wonder if the frequent setup/teardown of CephFS client sessions will slow things down but maybe it's not a significant problem. It'll be interesting to see some basic benchmarks to evaluate.

Reply:
Yes. We can maybe have a client session per fs volume, and use that to manage subvolume groups and subvolumes within the fs volume. Create and end session on creation and deletion of fs volumes?

Reply:
I think we would need some kind of connection pool a bit sooner (for background subvolume purge implementation) -- the purge queue worker needs to have a reference to the filesystem session for its activity (although it can init its own, but that will add to the overhead).

Reply:
The `libcephfs` client binding should be multithreaded (please confirm) so it should be possible to have one client per volume.